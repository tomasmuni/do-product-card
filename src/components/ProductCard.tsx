import React, { createContext } from 'react';

import { useProduct } from '../hooks/useProduct';
import { ProductContextProps, Product, OnChangeArgs, InitialValues, ProductCardHandlers } from '../interfaces/product.interfaces';

import styles from '../styles/styles.module.css'

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;



export interface Props {
    product: Product;
    // children?: React.ReactElement | React.ReactElement[];
    children: (args: ProductCardHandlers) => React.ReactElement
    className?: string;
    style?: React.CSSProperties;
    onChange?: ( args: OnChangeArgs ) => void;
    value?: number;
    initialValues?: InitialValues
}


export const ProductCard = ({ children, product, className, style, onChange, value , initialValues}: Props ) => {

    const { counter, increaseBy, maxCount, isMaxCountReached, reset } = useProduct({ onChange, product, value, initialValues });

    return (
        <Provider value={{
            counter,
            increaseBy,
            product,
            maxCount
        }}>
            <div 
                className={ `${ styles.productCard } ${ className }` }
                style={ style }
            >
                { children({
                    count: counter,
                    product,
                    isMaxCountReached,
                    increaseBy,
                    reset,
                    maxCount: initialValues?.count
                }) }
            </div>
        </Provider>
    )
}