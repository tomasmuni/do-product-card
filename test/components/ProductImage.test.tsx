import React from 'react'
import renderer from 'react-test-renderer'
import { ProductImage, ProductCard } from '../../src/components';
import { product1 } from '../data/products';



describe('ProductImage',() => {

    test("debe mostrar el componente con la imagen", () => {
        const wrapper = renderer.create(
            <ProductImage img={product1.img} className='custom-class'></ProductImage>
        )
        expect(wrapper.toJSON()).toMatchSnapshot();
    })

    test("debe mostrar el componente con la imagen", () => {
        const wrapper = renderer.create(
            <ProductCard product={product1}> 
            {
                () => (
                    <ProductImage img={product1.img}></ProductImage>
                )

            } 
            </ProductCard>
        )
        expect(wrapper.toJSON()).toMatchSnapshot();
    })

})