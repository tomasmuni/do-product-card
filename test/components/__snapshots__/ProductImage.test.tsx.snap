// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage debe mostrar el componente con la imagen 1`] = `
<img
  alt="Product"
  className="productImg custom-class"
  src="./coffee-mug.png"
/>
`;

exports[`ProductImage debe mostrar el componente con la imagen 2`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Product"
    className="productImg undefined"
    src="./coffee-mug.png"
  />
</div>
`;
